experiments/2017-09-21-rust-on-android.md

```diff
@@ -0,0 +1,244 @@
+
+We declared that we needed the `jni` crate, that means we need to include the crate in the `Cargo.toml` file. Open it up and add the following between the `[package]` and `[lib]` declarations.
+
+```
+[target.'cfg(target_os="android")'.dependencies]
+jni = { version = "0.5", default-features = false }
+```
+
+We are now ready to build our libraries. Unlike with iOS, there is no handy universal Android library that we can make so we have to create one for each of our target architectures. We can then create symlinks to them from the Android project. Navigate to your `cargo` directory and run the following commands:
+
+```
+cargo build --target aarch64-linux-android --release
+cargo build --target armv7-linux-androideabi --release
+cargo build --target i686-linux-android --release
+
+cd ../android/Greetings/app/src/main
+mkdir jniLibs
+mkdir jniLibs/arm64
+mkdir jniLibs/armeabi
+mkdir jniLibs/x86
+
+ln -s <path>/cargo/target/aarch64-linux-android/release/libgreetings.so jniLibs/arm64/libgreetings.so
+ln -s <path>/cargo/target/armv7-linux-androideabi/release/libgreetings.so jniLibs/armeabi/libgreetings.so
+ln -s <path>/cargo/target/i686-linux-android/release/libgreetings.so jniLibs/x86/libgreetings.so
+```
+
+Now, head back to Android Studio and open `GreetingsActivity.java`. We need to load our Rust library when the app starts, so add the following lines below the class declaration and before the `onCreate` method.
+
+```java
+   static {
+        System.loadLibrary("greetings");
+    }
+```
+
+This looks for a library called `greetings.so` inside the jniLibs directory and picks the correct one for the current architecture.
+
+Open `activity-greetings.xml`. In the `Component Tree` panel, highlight the `TextField` and open the `Properties` panel. Change the `ID` in the `Properties` panel to `greetingField`. This is how we are going to refer to it from our Activity.
+
+![Amend activity xml](assets/2017-09-21-rust-on-android/activity-greetings.png)
+
+Reopen `GreetingsActivity.java` and amend the `onCreate` method to call our greetings function and set the text on the `greetingField` `TextField` to the response value.
+
+```java
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_greetings);
+
+        RustGreetings g = new RustGreetings();
+        String r = g.sayHello("world");
+        ((TextView)findViewById(R.id.greetingField)).setText(r);
+    }
+```
+
+Build and run the app. If this is your first time in Android Studio, you may need to set up a simulator. When choosing/creating your simulator pick one with API 26. When the app starts, `Hello world` will be printed on your screen.
```

Comment:
So one last problem:
`Error:Could not list contents of '/Users/joe/projects/mozilla/greetings/android/Greetings/app/src/main/jniLibs/arm64/libgreetings.so'. Couldn't follow symbolic link.`

I solved it by doing a `cp` in place of the `ln -s` above

Reply:
Ah, yeah. See the above comment. I'll make this more obvious.